Marlin/src/module/stepper.cpp

```diff
@@ -2862,18 +2893,186 @@ hal_timer_t Stepper::block_phase_isr() {
+          count_direction.e = -count_direction.e;
+          DIR_WAIT_BEFORE();
+          E_APPLY_DIR(forward_e, false);
+          TERN_(FT_MOTION, last_set_direction = last_direction_bits);
+          DIR_WAIT_AFTER();
+        }
+      }
+    }
+    #if ENABLED(INPUT_SHAPING_E_SYNCH)
+      constexpr uint16_t IS_COMPENSATION_BUFFER_SIZE = 
+        (SMOOTH_LIN_ADV_HZ / SHAPING_MIN_FREQ / 2.0f + 0.5f);
+      
+      typedef struct {
+          xy_float_t buffer[IS_COMPENSATION_BUFFER_SIZE];
+          uint16_t index;
+      } DelayBuffer;
+
+      DelayBuffer delayBuffer;
+
+      void add_to_buffer(xy_float_t input) {
+        delayBuffer.buffer[delayBuffer.index] = input;
+        delayBuffer.index++;
+        if (delayBuffer.index == IS_COMPENSATION_BUFFER_SIZE) {
+          delayBuffer.index = 0;
+        }
+      }
+
+      xy_float_t lookback(shaping_time_t t /* in stepper timer ticks */) {
+        constexpr float ADV_TICKS_PER_STEPPER_TICKS = (float) SMOOTH_LIN_ADV_HZ / STEPPER_TIMER_RATE;
+        uint32_t delay_steps = t * ADV_TICKS_PER_STEPPER_TICKS  + 0.5f; // Convert time to steps
+        uint16_t past_i;
+        if (delay_steps>= IS_COMPENSATION_BUFFER_SIZE) {
+          // this means the buffer is too small. TODO: how to inform user?
+          past_i = delayBuffer.index;
+        }
+        else {
+          past_i = (delayBuffer.index + IS_COMPENSATION_BUFFER_SIZE - delay_steps);
+          if (past_i >= IS_COMPENSATION_BUFFER_SIZE) {
+            past_i -= IS_COMPENSATION_BUFFER_SIZE;
+          }
+        }
+        return delayBuffer.buffer[past_i];
+      }
+    #endif
+
+    float lookahead(uint32_t t) {
+      for (uint8_t i = 0; block_t * block = Planner::get_future_block(i); i++) {
+        if (block->is_sync()) continue;
+        if (t <= block->acceleration_time) {
+          if (!block->use_advance_lead) return 0.0f;
+          uint32_t rate = STEP_MULTIPLY(t, block->acceleration_rate) + block->initial_rate;
+          NOMORE(rate, block->nominal_rate);
+          return rate * block->e_step_ratio;
+        }
+        t -= block->acceleration_time;
+
+        if (t <= block->cruise_time) {
+          if (!block->use_advance_lead) return 0.0f;
+          return block->cruise_rate * block->e_step_ratio;
+        }
+        t -= block->cruise_time;
+
+        if (t <= block->deceleration_time) {
+          if (!block->use_advance_lead) return 0.0f;
+          uint32_t rate = STEP_MULTIPLY(t, block->acceleration_rate);
+          if (rate < block->cruise_rate) {
+            rate = block->cruise_rate - rate;
+            NOLESS(rate, block->final_rate);
+          }
+          else
+            rate = block->final_rate;
+          return rate * block->e_step_ratio;
+        }
+        t -= block->deceleration_time;
+      }
+      return 0.0f;
+    };
+    
+    hal_timer_t Stepper::smooth_lin_adv_isr() {
+      static float soothed_values[SMOOTH_LIN_ADV_EXP_ORDER] = {0};
+      float target_pressure = 0;
```

Comment:
The names of target_pressure and last_target_pressure confused me. They are not actually storing pressure but advanced steps. I think target_adv_steps or something similar would make more sense.

Reply:
You are right, renamed to target_adv_steps